extensions/amp-a4a/0.1/signature-verifier.js

```diff
@@ -85,3 +88,343 @@ export class ISignatureVerifier {
+      const keys = {};
+      const promise = this.fetchAndAddKeys_(keys, signingServiceName, null);
+      this.signers_[signingServiceName] = {promise, keys};
+    }
+  }
+
+  /**
+   * Extracts a cryptographic signature from `headers` and attempts to verify
+   * that it's the correct cryptographic signature for `creative`.
+   *
+   * As a precondition, `loadKeyset(signingServiceName)` must have already been
+   * called.
+   *
+   * @param {!ArrayBuffer} creative
+   * @param {!Headers} headers
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<!VerificationStatus>}
+   */
+  verify(creative, headers, lifecycleCallback) {
+    const signatureHeader = 'AMP-Fast-Fetch-Signature';
+    const signatureFormat =
+        /^([A-Za-z0-9._-]+):([A-Za-z0-9._-]+):([A-Za-z0-9+/]{341}[AQgw]==)$/;
+    if (!headers.has(signatureHeader)) {
+      return Promise.resolve(VerificationStatus.UNVERIFIED);
+    }
+    const headerValue = headers.get(signatureHeader);
+    const match = signatureFormat.exec(headerValue);
+    if (!match) {
+      // TODO(@taymonbeal, #9274): replace this with real error reporting
+      user().error('AMP-A4A', `Invalid signature header: ${headerValue}`);
+      return Promise.resolve(VerificationStatus.ERROR_SIGNATURE_MISMATCH);
+    }
+    return this.verifyCreativeAndSignature(
+        match[1], match[2], base64DecodeToBytes(match[3]), creative,
+        lifecycleCallback);
+  }
+
+  /**
+   * Verifies that `signature` is the correct cryptographic signature for
+   * `creative`, with the public key from the named signing service identified
+   * by `keypairId`.
+   *
+   * As a precondition, `loadKeyset(signingServiceName)` must have already been
+   * called.
+   *
+   * If the keyset for the named signing service was imported successfully but
+   * did not include a key for `keypairId`, this may be the result of a stale
+   * browser cache. To work around this, `keypairId` is added to the public key
+   * endpoint URL as a query parameter and the keyset is re-fetched. Other kinds
+   * of failures, including network connectivity failures, are not retried.
+   *
+   * @param {string} signingServiceName
+   * @param {string} keypairId
+   * @param {!Uint8Array} signature
+   * @param {!ArrayBuffer} creative
+   * @param {function(string, !Object)} lifecycleCallback called for each AMP
+   *     lifecycle event triggered during verification
+   * @return {!Promise<!VerificationStatus>}
+   * @visibleForTesting
+   */
+  verifyCreativeAndSignature(
+      signingServiceName, keypairId, signature, creative, lifecycleCallback) {
+    if (!this.signers_) {
+      // Web Cryptography isn't available.
+      return Promise.resolve(VerificationStatus.UNVERIFIED);
+    }
+    const signer = this.signers_[signingServiceName];
+    return signer.promise.then(success => {
```

Comment:
Should we limit the amount of time we're willing to wait for the public key?  Presumably there is some performance trade-off (perhaps add a TODO).